src/github.com/mozilla-services/pushgo/simplepush/worker.go

```diff
@@ -654,132 +714,187 @@ func (self *WorkerWS) Unregister(header *RequestHeader, message []byte) (err err
-			LogFields{"rid": self.logID, "cmd": "unregister"})
+	} else if w.logger.ShouldLog(DEBUG) {
+		w.logger.Debug("worker", "sending response",
+			LogFields{"rid": w.logID, "cmd": "unregister"})
 	}
-	self.WriteJSON(UnregisterReply{header.Type, 200, request.ChannelID})
-	self.metrics.Increment("updates.client.unregister")
+	w.WriteJSON(UnregisterReply{header.Type, 200, request.ChannelID})
+	w.metrics.Increment("updates.client.unregister")
 	return nil
 }
 
 // Dump any records associated with the UAID.
-func (self *WorkerWS) Flush(lastAccessed int64, channel string, version int64, data string) (err error) {
+func (w *WorkerWS) Flush(lastAccessed int64, channel string,
+	version int64, data string) (err error) {
+
 	// flush pending data back to Client
 	timer := timeNow()
-	logWarning := self.logger.ShouldLog(WARNING)
-	messageType := "notification"
-	uaid := self.UAID()
+	uaid := w.UAID()
 	defer func() {
 		now := timeNow()
-		if self.logger.ShouldLog(INFO) {
-			self.logger.Info("timer",
+		if w.logger.ShouldLog(INFO) {
+			w.logger.Info("worker",
 				"Client flush completed",
 				LogFields{"duration": strconv.FormatInt(int64(now.Sub(timer)), 10),
 					"uaid": uaid})
 		}
-		if err != nil || self.stopped() {
+		if err != nil || w.stopped() {
 			return
 		}
-		self.metrics.Timer("client.flush", now.Sub(timer))
+		w.metrics.Timer("client.flush", now.Sub(timer))
 	}()
 	if uaid == "" {
-		if logWarning {
-			self.logger.Warn("worker", "Undefined UAID for socket. Aborting.",
-				LogFields{"rid": self.logID})
+		if w.logger.ShouldLog(WARNING) {
+			w.logger.Warn("worker", "Undefined UAID for socket. Aborting.",
+				LogFields{"rid": w.logID})
 		}
 		// Have the server clean up records associated with this UAID.
 		// (Probably "none", but still good for housekeeping)
-		self.stop()
+		w.stop()
 		return nil
 	}
-	// Fetch the pending updates from #storage
-	var (
-		updates []Update
-		reply   *FlushReply
-	)
-	mod := false
-	// if we have a channel, don't flush. we can get them later in the ACK
-	if len(channel) == 0 {
-		var expired []string
-		if updates, expired, err = self.store.FetchAll(uaid, time.Unix(lastAccessed, 0)); err != nil {
-			if logWarning {
-				self.logger.Warn("worker", "Failed to flush Update to client.",
-					LogFields{"rid": self.logID, "uaid": uaid, "error": err.Error()})
-			}
-			return err
+	if len(channel) > 0 {
+		// if we have a channel, don't flush. we can get them later in the ACK
+		return w.flushUpdate(channel, version, data)
+	}
+	return w.flushPending(lastAccessed)
+}
+
+// flushUpdate writes an update containing chid, version, and data.
+func (w *WorkerWS) flushUpdate(chid string, version int64, data string) (
+	err error) {
+
+	uaid := w.UAID()
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
 		}
-		if len(updates) > 0 || len(expired) > 0 {
-			reply = &FlushReply{messageType, updates, expired}
+		pinger := w.app.PropPinger()
+		if pinger != nil {
+			pinger.Send(uaid, version, data)
+		}
```

Comment:
This sends a proprietary ping if `flushUpdate` panics. There is a test for this, but I don't think we've ever seen `WriteJSON` panic.